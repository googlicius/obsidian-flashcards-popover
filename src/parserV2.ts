import { CardType } from './enums';
import { generateRandomString } from './util/generateRandomString';

export interface ParseOptions {
	text: string;
	singlelineCardSeparator?: string;
	singlelineReversedCardSeparator?: string;
	multilineCardSeparator?: string;
	multilineReversedCardSeparator?: string;
	convertHighlightsToClozes?: boolean;
	convertBoldTextToClozes?: boolean;
	convertCurlyBracketsToClozes?: boolean;
	allTags?: string[];
}

export interface Card {
	type: CardType;
	text: string;
	lineNumber: number;
	tag?: string;
	sequenceId?: string;
	headings?: string[];
}

/**
 * Returns flashcards found in `text`
 *
 * @param options - Options for parsing flashcards
 * @returns An array of card objects
 */
export function parse({
	text,
	singlelineCardSeparator = '::',
	singlelineReversedCardSeparator = ':::',
	multilineCardSeparator = '?',
	multilineReversedCardSeparator = '??',
	convertHighlightsToClozes = true,
	convertBoldTextToClozes = true,
	convertCurlyBracketsToClozes = true,
	allTags = [],
}: ParseOptions): Card[] {
	let cardText = '';
	const cards: Card[] = [];
	let cardType: CardType | null = null;
	let lineNo = 0;
	let currentTag: string | undefined = undefined;
	let sequenceId: string | undefined = undefined;
	let tagRegex: RegExp | null = null;
	
	// Track headings and the line number where the current tag was found
	const headings: string[] = [];
	// Regular expression to match Markdown headings (# Heading)
	const headingRegex = /^(#{1,6})\s+(.+)$/;

	if (allTags.length > 0) {
		// Convert tags array to regex pattern
		const tagPattern = allTags
			.map((tag) => tag.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'))
			.join('|');

		// Create tag regex that matches tags only when they're on their own line
		tagRegex = new RegExp(`(?!.*::)(${tagPattern})\\b.*$`);
	}

	const blockRegex = /(@start|@end).*$/;

	const lines: string[] = text.replaceAll('\r\n', '\n').split('\n');
	for (let i = 0; i < lines.length; i++) {
		const currentLine = lines[i];
		const nextLine = lines[i + 1];

		// Check if the line is a heading and add it to the headings array
		const headingMatch = currentLine.match(headingRegex);
		if (headingMatch) {
			headings.push(headingMatch[2].trim());
		}

		if (currentLine.length === 0) {
			if (cardType) {
				// Create a card with headings if there's a current tag and headings
				cards.push({
					type: cardType,
					text: cardText,
					lineNumber: lineNo,
					tag: currentTag,
					sequenceId,
					headings: headings.length > 0 ? [...headings] : undefined,
				});
				cardType = null;
			}

			cardText = '';
			continue;
		} else if (
			tagRegex &&
			tagRegex.test(currentLine) &&
			(!nextLine || !['?', '??'].includes(nextLine.trim()))
		) {
			// Is a tag
			const match = currentLine.match(tagRegex) as string[];
			currentTag = match[1];
			// Reset headings when tag changes
			headings.length = 0;
		} else if (blockRegex.test(currentLine)) {
			// Is a block
			sequenceId = currentLine.startsWith('@start')
				? generateRandomString()
				: '';
		} else if (
			currentLine.startsWith('<!--') &&
			!currentLine.startsWith('<!--SR:')
		) {
			while (i + 1 < lines.length && !currentLine.includes('-->')) i++;
			i++;
			continue;
		}

		if (cardText.length > 0) {
			cardText += '\n';
		}
		cardText += currentLine.trimEnd();

		if (
			currentLine.includes(singlelineReversedCardSeparator) ||
			currentLine.includes(singlelineCardSeparator)
		) {
			cardType = lines[i].includes(singlelineReversedCardSeparator)
				? CardType.SingleLineReversed
				: CardType.SingleLineBasic;
			cardText = lines[i];
			lineNo = i;
			if (i + 1 < lines.length && lines[i + 1].startsWith('<!--SR:')) {
				cardText += '\n' + lines[i + 1];
				i++;
			}
			
			// Create a card with headings if there's a current tag and headings
			cards.push({
				type: cardType,
				text: cardText,
				lineNumber: lineNo,
				tag: currentTag,
				sequenceId,
				headings: headings.length > 0 ? [...headings] : undefined,
			});
			cardType = null;
			cardText = '';
		} else if (
			cardType === null &&
			((convertHighlightsToClozes && /==.*?==/gm.test(currentLine)) ||
				(convertBoldTextToClozes &&
					/\*\*.*?\*\*/gm.test(currentLine)) ||
				(convertCurlyBracketsToClozes && /{{.*?}}/gm.test(currentLine)))
		) {
			cardType = CardType.Cloze;
			lineNo = i;
		} else if (currentLine.trim() === multilineCardSeparator) {
			cardType = CardType.MultiLineBasic;
			lineNo = i;
		} else if (currentLine.trim() === multilineReversedCardSeparator) {
			cardType = CardType.MultiLineReversed;
			lineNo = i;
		} else if (
			currentLine.startsWith('```') ||
			currentLine.startsWith('~~~')
		) {
			const matches = currentLine.match(/`+|~+/);
			const codeBlockClose = matches && matches[0];
			if (codeBlockClose) {
				while (
					i + 1 < lines.length &&
					!lines[i + 1].startsWith(codeBlockClose)
				) {
					i++;
					cardText += '\n' + lines[i];
				}
				cardText += '\n' + codeBlockClose;
				i++;
			}
		}
	}

	if (cardType && cardText) {
		// Create a card with headings if there's a current tag and headings
		cards.push({
			type: cardType,
			text: cardText,
			lineNumber: lineNo,
			tag: currentTag,
			sequenceId,
			headings: headings.length > 0 ? [...headings] : undefined,
		});
	}

	return cards;
}
