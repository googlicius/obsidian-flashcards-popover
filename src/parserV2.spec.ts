import { CardType } from './enums';
import { parse } from './parserV2';

jest.mock('./util/generateRandomString', () => ({
	generateRandomString: jest.fn(() => 'faked'),
}));

test('Test parsing of single line basic cards', () => {
	expect(parse({ text: 'Question::Answer' })).toEqual([
		{
			type: CardType.SingleLineBasic,
			text: 'Question::Answer',
			lineNumber: 0,
		},
	]);
	expect(
		parse({ text: 'Question::Answer\n<!--SR:!2021-08-11,4,270-->' }),
	).toEqual([
		{
			type: CardType.SingleLineBasic,
			text: 'Question::Answer\n<!--SR:!2021-08-11,4,270-->',
			lineNumber: 0,
		},
	]);
	expect(
		parse({ text: 'Question::Answer <!--SR:2021-08-11,4,270-->' }),
	).toEqual([
		{
			type: CardType.SingleLineBasic,
			text: 'Question::Answer <!--SR:2021-08-11,4,270-->',
			lineNumber: 0,
		},
	]);
	expect(parse({ text: 'Some text before\nQuestion ::Answer' })).toEqual([
		{
			type: CardType.SingleLineBasic,
			text: 'Question ::Answer',
			lineNumber: 1,
		},
	]);
	expect(parse({ text: '#Title\n\nQ1::A1\nQ2:: A2' })).toEqual([
		{
			type: CardType.SingleLineBasic,
			text: 'Q1::A1',
			lineNumber: 2,
		},
		{
			type: CardType.SingleLineBasic,
			text: 'Q2:: A2',
			lineNumber: 3,
		},
	]);
	expect(parse({ text: '#flashcards/science Question ::Answer' })).toEqual([
		{
			type: CardType.SingleLineBasic,
			text: '#flashcards/science Question ::Answer',
			lineNumber: 0,
		},
	]);
});

test('Test parsing of single line reversed cards', () => {
	expect(parse({ text: 'Question:::Answer' })).toEqual([
		{
			type: CardType.SingleLineReversed,
			text: 'Question:::Answer',
			lineNumber: 0,
		},
	]);
	expect(
		parse({
			text: 'Some text before\nQuestion :::Answer',
		}),
	).toEqual([
		{
			type: CardType.SingleLineReversed,
			text: 'Question :::Answer',
			lineNumber: 1,
		},
	]);
	expect(parse({ text: '#Title\n\nQ1:::A1\nQ2::: A2' })).toEqual([
		{
			type: CardType.SingleLineReversed,
			text: 'Q1:::A1',
			lineNumber: 2,
		},
		{
			type: CardType.SingleLineReversed,
			text: 'Q2::: A2',
			lineNumber: 3,
		},
	]);
});

test('Test parsing of multi line basic cards', () => {
	expect(parse({ text: 'Question\n?\nAnswer' })).toEqual([
		{
			type: CardType.MultiLineBasic,
			text: 'Question\n?\nAnswer',
			lineNumber: 1,
		},
	]);
	expect(parse({ text: 'Question\n? \nAnswer' })).toEqual([
		{
			type: CardType.MultiLineBasic,
			text: 'Question\n?\nAnswer',
			lineNumber: 1,
		},
	]);
	expect(
		parse({
			text: 'Question\n?\nAnswer <!--SR:!2021-08-11,4,270-->',
		}),
	).toEqual([
		{
			type: CardType.MultiLineBasic,
			text: 'Question\n?\nAnswer <!--SR:!2021-08-11,4,270-->',
			lineNumber: 1,
		},
	]);
	expect(
		parse({
			text: 'Question\n?\nAnswer\n<!--SR:2021-08-11,4,270-->',
		}),
	).toEqual([
		{
			type: CardType.MultiLineBasic,
			text: 'Question\n?\nAnswer\n<!--SR:2021-08-11,4,270-->',
			lineNumber: 1,
		},
	]);
	expect(
		parse({
			text: 'Some text before\nQuestion\n?\nAnswer',
		}),
	).toEqual([
		{
			type: CardType.MultiLineBasic,
			text: 'Some text before\nQuestion\n?\nAnswer',
			lineNumber: 2,
		},
	]);
	expect(
		parse({
			text: 'Question\n?\nAnswer\nSome text after!',
		}),
	).toEqual([
		{
			type: CardType.MultiLineBasic,
			text: 'Question\n?\nAnswer\nSome text after!',
			lineNumber: 1,
		},
	]);
	expect(
		parse({
			text: '#Title\n\nLine0\nQ1\n?\nA1\nAnswerExtra\n\nQ2\n?\nA2',
		}),
	).toEqual([
		{
			type: CardType.MultiLineBasic,
			text: 'Line0\nQ1\n?\nA1\nAnswerExtra',
			lineNumber: 4,
		},
		{
			type: CardType.MultiLineBasic,
			text: 'Q2\n?\nA2',
			lineNumber: 9,
		},
	]);
	expect(
		parse({
			text: '#flashcards/tag-on-previous-line\nQuestion\n?\nAnswer',
		}),
	).toEqual([
		{
			type: CardType.MultiLineBasic,
			text: '#flashcards/tag-on-previous-line\nQuestion\n?\nAnswer',
			lineNumber: 2,
		},
	]);
});

test('Test parsing of multi line reversed cards', () => {
	expect(parse({ text: 'Question\n??\nAnswer' })).toEqual([
		{
			type: CardType.MultiLineReversed,
			text: 'Question\n??\nAnswer',
			lineNumber: 1,
		},
	]);
	expect(parse({ text: 'Some text before\nQuestion\n??\nAnswer' })).toEqual([
		{
			type: CardType.MultiLineReversed,
			text: 'Some text before\nQuestion\n??\nAnswer',
			lineNumber: 2,
		},
	]);
	expect(parse({ text: 'Question\n??\nAnswer\nSome text after!' })).toEqual([
		{
			type: CardType.MultiLineReversed,
			text: 'Question\n??\nAnswer\nSome text after!',
			lineNumber: 1,
		},
	]);
	expect(
		parse({
			text: '#Title\n\nLine0\nQ1\n??\nA1\nAnswerExtra\n\nQ2\n??\nA2',
		}),
	).toEqual([
		{
			type: CardType.MultiLineReversed,
			text: 'Line0\nQ1\n??\nA1\nAnswerExtra',
			lineNumber: 4,
		},
		{
			type: CardType.MultiLineReversed,
			text: 'Q2\n??\nA2',
			lineNumber: 9,
		},
	]);
});

test('Test parsing of cloze cards', () => {
	// ==highlights==
	expect(parse({ text: 'cloze ==deletion== test' })).toEqual([
		{
			type: CardType.Cloze,
			text: 'cloze ==deletion== test',
			lineNumber: 0,
		},
	]);
	expect(
		parse({ text: 'cloze ==deletion== test\n<!--SR:2021-08-11,4,270-->' }),
	).toEqual([
		{
			type: CardType.Cloze,
			text: 'cloze ==deletion== test\n<!--SR:2021-08-11,4,270-->',
			lineNumber: 0,
		},
	]);
	expect(
		parse({ text: 'cloze ==deletion== test <!--SR:2021-08-11,4,270-->' }),
	).toEqual([
		{
			type: CardType.Cloze,
			text: 'cloze ==deletion== test <!--SR:2021-08-11,4,270-->',
			lineNumber: 0,
		},
	]);
	expect(parse({ text: '==this== is a ==deletion==\n' })).toEqual([
		{
			type: CardType.Cloze,
			text: '==this== is a ==deletion==',
			lineNumber: 0,
		},
	]);
	expect(
		parse({
			text:
				'some text before\n\na deletion on\nsuch ==wow==\n\n' +
				'many text\nsuch surprise ==wow== more ==text==\nsome text after\n\nHmm',
		}),
	).toEqual([
		{
			type: CardType.Cloze,
			text: 'a deletion on\nsuch ==wow==',
			lineNumber: 3,
		},
		{
			type: CardType.Cloze,
			text: 'many text\nsuch surprise ==wow== more ==text==\nsome text after',
			lineNumber: 6,
		},
	]);
	expect(parse({ text: 'srdf ==' })).toEqual([]);
	expect(parse({ text: 'lorem ipsum ==p\ndolor won==' })).toEqual([]);
	expect(parse({ text: 'lorem ipsum ==dolor won=' })).toEqual([]);
	// ==highlights== turned off
	expect(
		parse({
			text: 'cloze ==deletion== test',
			convertHighlightsToClozes: false,
			convertBoldTextToClozes: true,
		}),
	).toEqual([]);

	// **bolded**
	expect(parse({ text: 'cloze **deletion** test' })).toEqual([
		{
			type: CardType.Cloze,
			text: 'cloze **deletion** test',
			lineNumber: 0,
		},
	]);
	expect(
		parse({ text: 'cloze **deletion** test\n<!--SR:2021-08-11,4,270-->' }),
	).toEqual([
		{
			type: CardType.Cloze,
			text: 'cloze **deletion** test\n<!--SR:2021-08-11,4,270-->',
			lineNumber: 0,
		},
	]);
	expect(
		parse({ text: 'cloze **deletion** test <!--SR:2021-08-11,4,270-->' }),
	).toEqual([
		{
			type: CardType.Cloze,
			text: 'cloze **deletion** test <!--SR:2021-08-11,4,270-->',
			lineNumber: 0,
		},
	]);
	expect(parse({ text: '**this** is a **deletion**\n' })).toEqual([
		{
			type: CardType.Cloze,
			text: '**this** is a **deletion**',
			lineNumber: 0,
		},
	]);
	expect(
		parse({
			text:
				'some text before\n\na deletion on\nsuch **wow**\n\n' +
				'many text\nsuch surprise **wow** more **text**\nsome text after\n\nHmm',
		}),
	).toEqual([
		{
			type: CardType.Cloze,
			text: 'a deletion on\nsuch **wow**',
			lineNumber: 3,
		},
		{
			type: CardType.Cloze,
			text: 'many text\nsuch surprise **wow** more **text**\nsome text after',
			lineNumber: 6,
		},
	]);
	expect(parse({ text: 'srdf **' })).toEqual([]);
	expect(parse({ text: 'lorem ipsum **p\ndolor won**' })).toEqual([]);
	expect(parse({ text: 'lorem ipsum **dolor won*' })).toEqual([]);
	// **bolded** turned off
	expect(
		parse({
			text: 'cloze **deletion** test',
			convertHighlightsToClozes: true,
			convertBoldTextToClozes: false,
		}),
	).toEqual([]);

	// both
	expect(
		parse({ text: 'cloze **deletion** test ==another deletion==!' }),
	).toEqual([
		{
			type: CardType.Cloze,
			text: 'cloze **deletion** test ==another deletion==!',
			lineNumber: 0,
		},
	]);
});

test('Test parsing of a mix of card types', () => {
	expect(
		parse({
			text:
				'# Lorem Ipsum\n\nLorem ipsum dolor ==sit amet==, consectetur ==adipiscing== elit.\n' +
				'Duis magna arcu, eleifend rhoncus ==euismod non,==\nlaoreet vitae enim.\n\n' +
				'Fusce placerat::velit in pharetra gravida\n\n' +
				'Donec dapibus ullamcorper aliquam.\n??\nDonec dapibus ullamcorper aliquam.\n<!--SR:2021-08-11,4,270-->',
		}),
	).toEqual([
		{
			type: CardType.Cloze,
			text:
				'Lorem ipsum dolor ==sit amet==, consectetur ==adipiscing== elit.\n' +
				'Duis magna arcu, eleifend rhoncus ==euismod non,==\n' +
				'laoreet vitae enim.',
			lineNumber: 2,
			headings: ['Lorem Ipsum'],
		},
		{
			type: CardType.SingleLineBasic,
			text: 'Fusce placerat::velit in pharetra gravida',
			lineNumber: 6,
			headings: ['Lorem Ipsum'],
		},
		{
			type: CardType.MultiLineReversed,
			text: 'Donec dapibus ullamcorper aliquam.\n??\nDonec dapibus ullamcorper aliquam.\n<!--SR:2021-08-11,4,270-->',
			lineNumber: 9,
			headings: ['Lorem Ipsum'],
		},
	]);
});

test('Test codeblocks', () => {
	// no blank lines
	expect(
		parse({
			text:
				'How do you ... Python?\n?\n' +
				"```\nprint('Hello World!')\nprint('Howdy?')\nlambda x: x[0]\n```",
		}),
	).toEqual([
		{
			type: CardType.MultiLineBasic,
			text:
				'How do you ... Python?\n?\n' +
				"```\nprint('Hello World!')\nprint('Howdy?')\nlambda x: x[0]\n```",
			lineNumber: 1,
		},
	]);

	// with blank lines
	expect(
		parse({
			text:
				'How do you ... Python?\n?\n' +
				"```\nprint('Hello World!')\n\n\nprint('Howdy?')\n\nlambda x: x[0]\n```",
		}),
	).toEqual([
		{
			type: CardType.MultiLineBasic,
			text:
				'How do you ... Python?\n?\n' +
				"```\nprint('Hello World!')\n\n\nprint('Howdy?')\n\nlambda x: x[0]\n```",
			lineNumber: 1,
		},
	]);

	// general Markdown syntax
	expect(
		parse({
			text:
				'Nested Markdown?\n?\n' +
				'````ad-note\n\n' +
				'```git\n' +
				"+ print('hello')\n" +
				"- print('world')\n" +
				'```\n\n' +
				'~~~python\n' +
				"print('hello world')\n" +
				'~~~\n' +
				'````',
		}),
	).toEqual([
		{
			type: CardType.MultiLineBasic,
			text:
				'Nested Markdown?\n?\n' +
				'````ad-note\n\n' +
				'```git\n' +
				"+ print('hello')\n" +
				"- print('world')\n" +
				'```\n\n' +
				'~~~python\n' +
				"print('hello world')\n" +
				'~~~\n' +
				'````',
			lineNumber: 1,
		},
	]);
});

test.skip('Test not parsing cards in HTML comments', () => {
	expect(
		parse({
			text: '<!--\nQuestion\n?\nAnswer <!--SR:!2021-08-11,4,270-->\n-->',
		}),
	).toEqual([]);
	expect(
		parse({
			text: '<!--\nQuestion\n?\nAnswer <!--SR:!2021-08-11,4,270-->\n\n<!--cloze ==deletion== test-->-->',
		}),
	).toEqual([]);
	expect(parse({ text: '<!--cloze ==deletion== test-->' })).toEqual([]);
	expect(parse({ text: '<!--cloze **deletion** test-->' })).toEqual([]);
});

test('Test cards in a sequence', () => {
	const text = '@start\n\nQ1 :: A1\n\nQ2 :: A2\n\n@end\nQ3::A3';
	expect(parse({ text })).toMatchSnapshot();
});

test('test section tag', () => {
	const text =
		'#flashcards/math\n\nQ1 :: A1\n\n#flashcards/chem\nQ2 :: A2\n\nQ3::A3';
	expect(
		parse({
			text,
			allTags: ['#flashcards/math', '#flashcards/chem'],
		}),
	).toMatchSnapshot();
});

test('Test cards with headings', () => {
	expect(
		parse({
			text: '# Heading 0\nSome text\n#test_1\n\n# Heading 1\n\n# Heading 2\n\nQ1::A1\n#test_2\nQ2::A2\n#test_3\n# Heading A\n\nQ3::A3',
			allTags: ['#test_1', '#test_2','#test_3'],
		}),
	).toMatchObject([
		{
			lineNumber: 8,
			tag: '#test_1',
			text: 'Q1::A1',
			type: 0,
			headings: ['Heading 1', 'Heading 2'],
		},
		{
			lineNumber: 10,
			tag: '#test_2',
			text: 'Q2::A2',
			type: 0,
		},
		{
			lineNumber: 14,
			tag: '#test_3',
			text: 'Q3::A3',
			type: 0,
			headings: ['Heading A']
		},
	]);
});
