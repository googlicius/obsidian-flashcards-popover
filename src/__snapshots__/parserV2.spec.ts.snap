// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test cards in a sequence 1`] = `
[
  {
    "headings": undefined,
    "lineNumber": 2,
    "sequenceId": "faked",
    "tag": undefined,
    "text": "Q1 :: A1",
    "type": 0,
  },
  {
    "headings": undefined,
    "lineNumber": 4,
    "sequenceId": "faked",
    "tag": undefined,
    "text": "Q2 :: A2",
    "type": 0,
  },
  {
    "headings": undefined,
    "lineNumber": 7,
    "sequenceId": "",
    "tag": undefined,
    "text": "Q3::A3",
    "type": 0,
  },
]
`;

exports[`test section tag 1`] = `
[
  {
    "headings": undefined,
    "lineNumber": 2,
    "sequenceId": undefined,
    "tag": "#flashcards/math",
    "text": "Q1 :: A1",
    "type": 0,
  },
  {
    "headings": undefined,
    "lineNumber": 5,
    "sequenceId": undefined,
    "tag": "#flashcards/chem",
    "text": "Q2 :: A2",
    "type": 0,
  },
  {
    "headings": undefined,
    "lineNumber": 7,
    "sequenceId": undefined,
    "tag": "#flashcards/chem",
    "text": "Q3::A3",
    "type": 0,
  },
]
`;
